R/lfhtcID.R

```diff
@@ -161,108 +120,320 @@ getPossibleYZ <- function(graph, L, allowedForY, allowedForZ) {
+#'   vector containing the nodes L otherwise it is empty.}
+#' }
+lfhtcID <- function(graph){
+
+  # Check the graph
+  latentDigraphHasSimpleNumbering(graph)
+
+  unsolvedParents <- lapply(graph$observedNodes(), graph$observedParents)
+  solvedParents <- rep(list(numeric(0)), length(graph$observedNodes()))
+  activeFroms <- rep(list(numeric(0)), length(graph$observedNodes()))
+  Zs <- rep(list(numeric(0)), length(graph$observedNodes()))
+  Ls <- rep(list(numeric(0)), length(graph$observedNodes()))
+  identifier <- createLFIdentifierBaseCase(graph)
+
+  changeFlag <- T
+  while (changeFlag) {
+    idResult <- lfhtcIdentifyStep(graph, unsolvedParents, solvedParents, activeFroms, Zs, Ls, identifier)
+    changeFlag <- (nrow(idResult$identifiedEdges) != 0)
+    unsolvedParents <- idResult$unsolvedParents
+    solvedParents <- idResult$solvedParents
+    activeFroms <- idResult$activeFroms
+    Zs <- idResult$Zs
+    Ls <- idResult$Ls
+    identifier <- idResult$identifier
+  }
+
+  result <- list()
+  class(result) <- "LfhtcIDResult"
+  result$solvedParents <- solvedParents
+  result$unsolvedParents <- unsolvedParents
+  result$identifier <- identifier
+  result$graph <- graph
+  result$activeFroms <- activeFroms
+  result$Zs <- Zs
+  result$Ls <- Ls
+  result$call <- match.call()
+  return(result)
+}
+
+
+
+#' Prints a LfhtcIDResult object
+#'
+#' Prints a LfhtcIDResult object as returned by
+#' \code{\link{lfhtcID}}. Invisibly returns its argument via
+#' \code{\link{invisible}(x)} as most print functions do.
+#'
+#' @export
+#'
+#' @param x the LfhtcIDResult object
+#' @param ... optional parameters, currently unused.
+print.LfhtcIDResult <- function(x, ...) {
+  cat("Call: ")
+  print(x$call)
+
+  observedNodes <- x$graph$observedNodes()
+  nObserved <- length(observedNodes)
+  latentNodes <- x$graph$latentNodes()
+  nLatent <- length(latentNodes)
+  solvedParents <- x$solvedParents
+
+  cat(paste("\nLatent Digraph Info\n"))
+  cat(paste("# observed nodes:", nObserved, "\n"))
+  cat(paste("# latent nodes:", nLatent, "\n"))
+  cat(paste("# total nr. of edges between observed nodes:", sum(x$graph$L()[observedNodes,]), "\n"))
+
+  cat(paste("\nGeneric Identifiability Summary\n"))
+  cat(paste("# nr. of edges between observed nodes shown gen. identifiable:", length(unlist(solvedParents)),
+            "\n"))
+
+  cat("# gen. identifiable edges: ")
+  edges <- character(min(length(unlist(solvedParents))/2, 11))
+  k <- 0
+  for (i in 1:nObserved) {
+    if (length(solvedParents[[i]]) != 0) {
+      for (j in solvedParents[[i]]) {
+        k <- k + 1
+        edges[k] <- paste(j, "->", i, sep = "")
+
+        if (k == 10) {
+          edges[11] <- "..."
+          break
+        }
+      }
+      if (k == 10) {
+        break
+      }
+    }
+  }
+  if (length(edges) == 0) {
+    cat("None\n")
+  } else {
+    cat(paste(paste(edges, collapse = ", "), "\n"))
+  }
+
+  invisible(x)
 }
```

Comment:
Always good practice to end files in a new line (I believe this is my fault though).